ql/src/java/org/apache/hadoop/hive/ql/optimizer/correlation/ReduceSinkDeDuplicationUtils.java

```diff
@@ -181,6 +183,23 @@ public static boolean merge(HiveConf hiveConf, ReduceSinkOperator cRS, ReduceSin
         TableDesc keyTable = PlanUtils.getReduceKeyTableDesc(new ArrayList<FieldSchema>(), pRS
             .getConf().getOrder(), pRS.getConf().getNullOrder());
         pRS.getConf().setKeySerializeInfo(keyTable);
+      } else if (cRS.getConf().getKeyCols() != null && cRS.getConf().getKeyCols().size() > 0) {
```

Comment:
don't we need any conditional on `pRS` here?

Reply:
setNumDistributionKeys is a subset of keycols, We enters this conditions only when NumDistributionKeys of pRS is null or <= 0. Hence checking for pRS doesn't make sense here since we anyhow want to go with cRS.